File: javascript/socketd/src/socketd/transport/core/ChannelAssistant.ts
import type {Frame} from "./Frame";
import {SocketAddress} from "./SocketAddress";

/**
 * 通道助理
 *
 * @author noear
 * @since 2.0
 */
export interface ChannelAssistant<T> {
    /**
     * 写入
     *
     * @param target 目标
     * @param frame  帧
     */
    write(target: T, frame: Frame);

    /**
     * 是否有效
     */
    isValid(target: T): boolean;

    /**
     * 关闭
     */
    close(target: T);

    /**
     * 获取远程地址
     */
    getRemoteAddress(target: T): SocketAddress | null;

    /**
     * 获取本地地址
     */
    getLocalAddress(target: T): SocketAddress | null;
}

File: python/socketd/socketd/transport/core/ChannelAssistant.py
import typing
from typing import Any
from abc import ABC, abstractmethod
from socketd.transport.core.Frame import Frame

S = typing.TypeVar("S")


class ChannelAssistant(typing.Generic[S], ABC):
    @abstractmethod
    async def write(self, target: Any, frame: Frame) -> None:
        ...

    @abstractmethod
    def read(self, target: Any) -> Frame: ...

    @abstractmethod
    def is_valid(self, target: Any) -> bool:
        ...

    @abstractmethod
    async def close(self, target: Any) -> None:
        ...

    @abstractmethod
    def get_remote_address(self, target: Any) -> str:
        ...

    @abstractmethod
    def get_local_address(self, target: Any) -> str:
        ...
